---
title: "Lab2"
output: html_document
editor_options: 
  chunk_output_type: console
---

# More R-basics

## Indexing data structures

When we want to extract particular values or subsets of a data structure (vector, matrix or data frame) in R, we call that "indexing".

To index a vector, put in [] the number corresponding to the position you want to extract:

```{r}
myvec <- c("a","b","c","d","e")
myvec[3]
```

If you want to extract a subset of a vector (more than one position), put a vector of integers corresponding to those positions in []:

```{r}
myvec[1:3]
myvec[c(1,3,5)]
```

To index 2D data structures such as matrices and data frames, we also use square brackets. The difference is that there is a comma in there, and the value(s) before the comma represent line numbers, and column numbers are after the comma.

```{r}
mymat <- matrix(c("a","b","c","d","e","f"), ncol=3)
mymat
#Rows, columns
mymat[1,2]
mymat[1,c(1,3)]
mymat[1,]
```

## Vectorized operations

Many operations in R are vectorized. Let's look at some examples of what this means:

```{r}
numvec <- 1:10
numvec
# simple vectorized math
numvec+1
numvec2 <- numvec*10
numvec2
# math combining vectors of same length
numvec2 - numvec
(numvec2 - numvec)^2
```

# Chi-square test for HWE

Last week we explored data from harpy populations in Brazil and in Mexico.
Here's a summary of the data:

We sampled 90 individuals in each location, genotyped a single locus in their genome, and observed the following genotypes:

*Brazil:* 44 AA, 38 Aa, 8 aa 
*Mexico:* 59 AA, 8 Aa, 23 aa

In both populations the frequency of allele *A* was 0.7 and frequency of allele *a* was 0.3, but we can see their genotype frequencies are very different.
We calculated expected genotype frequencies under HWE using a function similar to the following:

```{r}
hwe_freqs <- function(f_A){
  # calculate the other allele frequency
  f_a <- 1-f_A
  # calculate the 3 genotype frequencies
  f_AA <- f_A^2
  f_Aa <- 2*f_A*f_a
  f_aa <- f_a^2
  # return results as a VECTOR
  return(c("f_AA"=f_AA, "f_Aa"=f_Aa, "f_aa"=f_aa))
}
```

We then compared expected and observed genotype frequencies within each population by eye.
Now let's compare observed genotype frequencies to the expected under HWE in each of those populations more formally.

**Q1** Write a function that  takes a vector with observed genotype counts in the order AA, Aa and aa, as an input and calculates the chi-square statistic. Return the chi-squared statistic value.

Example:
calc_chi(c(44,38,8))
[1] 0.002519526

**Hint** You can write a function that takes a vector with observed genotype counts as an input. Calculate the genotype frequencies. Obtain the allele frequencies. Call the function *hwe_freqs* with the *f_A* allele frequency for calculating the expected frequencies. Calculate the $\chi^2$.

You do not have to follow the instructions of this hint necessarily.

```{r}
# CODE YOUR ANSWER HERE


```


**Q2** Use the function from Q1 to calculate the chi-square statistic for the harpy populations. The null hypothesis is that genotype frequencies follow HWE in each of the populations.

```{r}
# CODE YOUR ANSWER HERE


```

The chi-square test for HWE has 1 degre of freedom (d.f.). The code below generates a plot of the chi-square distribution with 1 d.f. and looks at some of its characteristics:

```{r}
# plot chi-square distribution with 1 d.f.
curve(dchisq(x,df=1), from=0, to=30,
      ylab="Probability density", xlab="chi-square")

# critical value at 5% significance value is 3.841. 
qchisq(p = 0.95, df=1)
# i.e. 95% of the chi-square values are below that.
pchisq(q=3.841, df=1)

# include red vertical line at critical value of 3.841
abline(v=3.841, col="red")
```

**Q3a** Calculate the p-value for the chi-square statistics (use lower.tail=FALSE)
```{r}
# CODE YOUR ANSWER HERE


```

**Q3b** Based on your results from Q2, would you say the Brazilian population is under HWE?
**A:**

**Q3c** What about the Mexican population?

**A:**

# Wright-Fisher model of genetic drift

A Wright-Fisher population is a model that shows the effect of genetic drift on changes in allele frequency through time.

In a Wright-Fisher population, the frequency of an allele will change from one generation to the next only due to stochastic sampling effects.
Every individual has on average 1 offspring, but some individuals will have 0 or more than 1 offspring. This probability is given by the binomial distribution. 

The probability of an individual *j* in generation t having *x* offspring in the next generation is given by the binomial distribution. This probability depends on the population size (*2N*), and assumes that every individual in theneration *t* has the same probability (*1/2N*) of being a parent of an individual in generation *t+1*. In R, we can calculate that probability with the function *dbinom*, as follows:

```{r}
# Example: probability that an individual has 0 offspring in a population of 30 individuals
twoN <- 30
dbinom(x = 0, size = twoN, prob = 1/twoN)
# 0 success in 30 trials with probability of success 1/30 in each trial
```

Now, let's look at the effect of drift on allele frequencies.

The frequency of an allele is not expected to change in a particular direction between generations. But it will change randomly due to sampling effects like the random deviations in number of offspring per individual we showed above. 

**Example**

We observe 10 copies of allele *A* in the current generation, in a population of 20 individuals. The chance that we observe 15 copies of *A* in the next generation is:

```{r}
next_gen_allele_copies <- 15 # number of successes
twoN <- 20 # number of trials
current_af <- 10/twoN # probability of success

dbinom(x = next_gen_allele_copies, size = twoN, prob = current_af)
```

**Q4** Find the probability that fA increases from 0.5 to 0.6 in a single generation in a population of 2N = 20. What about 2N = 200?

```{r}

```


**Q5** Compare your answers from Q4. Briefly explain what this demonstrates about the relationship between the force of drift and population size.

**A**


## Wright-Fisher simulations

Use the driftR simulator to convince yourself of the result you describe in Q5. For each of the following conditions, run a simulation starting with *fA=0.5* for *100 populations* for *250 generations*:
*Population size = 10*
*Population size = 100*
*Population size = 1000*

(We are simulating Wright-Fisher populations, so there should be *no mutation, no migration and fitness of all genotypes should be 1*).

```{r}
library(shiny)
runGitHub(username="cjbattey",repo="driftR")
```


**Q6** For each case, what proportion of populations either fixed or lost allele A? 

**A**
Population size = 10,    fixed =      , lost =  
Population size = 100,   fixed =      , lost = 
Population size = 1000,  fixed =      , lost = 

**Q7** All other things equal, would you expect more genetic variability in a large or small population? Why?

**A**

